include/execpools/thread_pool_base.hpp

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright (c) 2023 Ben FrantzDale
+ * Copyright (c) 2021-2023 Facebook, Inc. and its affiliates.
+ * Copyright (c) 2021-2024 NVIDIA Corporation
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <exec/static_thread_pool.hpp>
+
+namespace execpools {
+  //! This is a P2300-style thread pool wrapping tbb::task_arena, which its docs describe as "A class that represents an
+  //! explicit, user-managed task scheduler arena."
+  //! Once set up, a tbb::task_arena has
+  //! * template<F> void enqueue(F &&f)
+  //! and
+  //! * template<F> auto execute(F &&f) -> decltype(f())
+  //!
+  //! See https://spec.oneapi.io/versions/1.0-rev-3/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html
+  using namespace stdexec::tags;
+
+  template <class PoolType, class ReceiverId>
+  struct operation {
+    using Receiver = stdexec::__t<ReceiverId>;
+    struct __t;
```

Comment:
Aside: I keep seeing this pattern, along with `stdexec::__t<Foo>`. I see what it's doing, but it's not clear to me why, other than that `__t<Foo>` is nicer than `typename Foo::__t`. If someone could give an explanation, I'd gladly add a doxygen comment to
```cpp
  template <class _Tp>
  using __t = typename _Tp::__t;
```
in `__meta.hpp`.

Reply:
> other than that `__t<Foo>` is nicer than `typename Foo::__t`.

that's it, really. i don't like littering my code with `typename` everywhere. not a hill to die on.

Reply:
Makes sense. With that in mind, why is it seemingly everywhere? Like
```cpp
    class static_thread_pool_ {
      template <class ReceiverId>
      struct operation {
        using Receiver = stdexec::__t<ReceiverId>;
        class __t;
      };
```
That's saying that `static_thread_pool_`'s `operation<ReceiverId>` struct has a `Receiver`, which it gets from the `ReceiverId`, and then has a `__t` class which is the operation state? That gets used later to do `template <class Receiver> using operation_t = stdexec::__t<operation<stdexec::__id<Receiver>>>;` Why that indirection there? (If you can explain it, I'll add `//!` to relevant places!) To me it feels like a round-about way of saying that `operation_t<Receiver>` is the operation-state type: You get the `__id` of the receiver, get `operation<that>` which basically says `using Receiver = __t<__id<Receiver>>;`...

I'm sure the indirection is there for a reason, I just don't see it. Is there some more-universal pattern going on that this is just a simple case of?

Reply:
If I don't misunderstand the question - so feel free to correct me if I do  - which is why the indirection is made with the `__t` class and what is the reason behind this indirection. Then for me this [doc](https://github.com/NVIDIA/stdexec/blob/519f931ef9d44ca6cc8b2c8a3b3cb760a3523d4f/MAINTAINERS.md#class-template-parameters) pretty much explains it.
Briefly as I understand: It makes the class type less tight coupled from its template arguments (I.e.: It depends on the Id as argument not on the specific type)

Is it correct @ericniebler? Please correct me when it's wrong.

On the other hand I'm curious whether c++-modules will make changes on it one day or not, but it is just a personal curiosity.

Reply:
Aah. If there are no objections I’ll add a comment to `__t` with breadcrumbs pointing to that doc. Thank you!